hazelcast/src/main/java/com/hazelcast/internal/dynamicclassloading/impl/ClassLocator.java

```diff
@@ -0,0 +1,207 @@
+import com.hazelcast.internal.dynamicclassloading.impl.operation.ClassDataFinderOperation;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.IOUtil;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.OperationService;
+
+import java.io.Closeable;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+/**
+ * Provides classes to a local member.
+ *
+ * It's called by {@link com.hazelcast.internal.dynamicclassloading.DynamicClassLoader} when a class
+ * is not found on a local classpath.
+ *
+ * The current implementation can consult a cache and when the class is not found then it consults
+ * remote members.
+ *
+ */
+public final class ClassLocator {
+    private final ConcurrentMap<String, ClassSource> classSourceMap;
+    private final ClassLoader parent;
+    private final Filter<String> classNameFilter;
+    private final Filter<Member> memberFilter;
+    private final DynamicClassloadingConfig.ClassCacheMode classCacheMode;
+    private final NodeEngine nodeEngine;
+    private final ClassloadingMutexProvider mutexFactory = new ClassloadingMutexProvider();
+    private final ILogger logger;
+
+    public ClassLocator(ConcurrentMap<String, ClassSource> classSourceMap,
+                        ClassLoader parent, Filter<String> classNameFilter, Filter<Member> memberFilter,
+                        DynamicClassloadingConfig.ClassCacheMode classCacheMode, NodeEngine nodeEngine) {
+        this.classSourceMap = classSourceMap;
+        this.parent = parent;
+        this.classNameFilter = classNameFilter;
+        this.memberFilter = memberFilter;
+        this.classCacheMode = classCacheMode;
+        this.nodeEngine = nodeEngine;
+        this.logger = nodeEngine.getLogger(ClassLocator.class);
+    }
+
+    public static void onStartDeserialization() {
+        ThreadLocalClassCache.onStartDeserialization();
+    }
+
+    public static void onFinishDeserialization() {
+        ThreadLocalClassCache.onFinishDeserialization();
+    }
+
+    public Class<?> handleClassNotFoundException(String name)
+            throws ClassNotFoundException {
+        if (!classNameFilter.accept(name)) {
+            throw new ClassNotFoundException("Class " + name + " is not allowed to be loaded from other members.");
+        }
+        Class<?> clazz = tryToGetClassFromLocalCache(name);
+        if (clazz != null) {
+            return clazz;
+        }
+        return tryToGetClassFromRemote(name);
+    }
+
+    private Class<?> tryToGetClassFromRemote(String name) throws ClassNotFoundException {
+        //we need to acquire a classloading lock before defining a class
+        //Java 7+ can use locks with per-class granularity while Java 6
+        //has to use a single lock.
+        //mutexFactory abstract these differences away
+        Closeable classMutex = mutexFactory.getMutexForClass(name);
+        try {
+            synchronized (classMutex) {
```

Comment:
It's a mutex per classname (in Java 7+)
I guess you could come-up with a scenario where dead-lock can occur, but can't think of any realistic one. It's all circle dependencies where depended classes are spread over multiple members.